### ISSUE TITLE
Enemy patrol groups generate waypoint at [0,0,0] coordinates

### ISSUE THREAD

**Issue by GoldJohnKing** [Vdauphin/HeartsAndMinds#935]
**Arma 3 Version:** `1.98` (stable)
**CBA Version:** `3.15.1` (stable)
**ACE3 Version:** `3.13.3` (stable)
**RHS USAF Version:** `0.5.2` (stable)
**Hearts and minds Version:** `master_daily` (c8d0e82, migrate to Kujari and slightly modified)

**Mods:**
```
- CBA_A3
- ace
- RHS USAF
- RHS AFRF
- RHS GREF
- RHS SAF
- CUP Terrains - Core
- Kujari
- USAF MOD Main Pack
- USAF MOD Utility pack
- USAF MOD Fighters Pack
- JSRS SOUNDMOD
```

**Description:**
- Enemy patrol groups spawned by `btc_fnc_patrol_init` would generate waypoints at [0,0,0] coordinates.
- They should generate all waypoints towards their destination city, but actually, they first generate waypoint to [0,0,0], then to their destination city.
- But they do generate final waypoint to their destination city correctly.
- So I think there might be something misbehavior in `btc_fnc_patrol_init`, `btc_fnc_patrol_addWP` and maybe `btc_fnc_patrol_WPCheck` (because this function also called `btc_fnc_patrol_init`).

**Steps to reproduce:**
- Not sure, I only noticed this randomly.

**Screenshot:**
![107410_20200807103224_1](https://user-images.githubusercontent.com/26923626/89603122-f11d1680-d89a-11ea-8ddf-9943e08d7294.png)

**Comment by Vdauphin** [comment]
Please trow the .rtp .pbo ...

**Comment by GoldJohnKing** [comment]
[=BTC=co@30_Hearts_and_Minds.tem_kujari.zip](https://github.com/Vdauphin/HeartsAndMinds/files/5040174/BTC.co%4030_Hearts_and_Minds.tem_kujari.zip)

[Arma3_x64_2020-08-07_14-03-03.zip](https://github.com/Vdauphin/HeartsAndMinds/files/5040194/Arma3_x64_2020-08-07_14-03-03.zip)

**Comment by Vdauphin** [comment]
Please trow the real mod list and rpt link to the pbo where the bug occur

This pbo require SSS_module which is not your rpt

**Comment by Vdauphin** [comment]
need ffaa also apparently

**Comment by GoldJohnKing** [comment]
Ah, sorry, the pbo is wrong...

This is the real rpt that the error occoured, I tried to avoid any AI related mods (SSS is a AI support module called Simplex Support Service) and other unrelated mods, then run a clean mission in this rpt.

I'll upload a exact pbo as soon as possible.

And I found another detail on the issue... The issue groups are with a "loop waypoint", and they do not despawn as city deactivate.

See: 

![107410_20200807170203_1](https://user-images.githubusercontent.com/26923626/89630110-758a8c00-d8d1-11ea-8fa9-2e46824b3451.png)

**Comment by Vdauphin** [comment]
need USAF Mod also

**Comment by GoldJohnKing** [comment]
USAF should be already included in the rpt file.

Here's the correct mission pbo.

[H&M_Kujari.tem_kujari.zip](https://github.com/Vdauphin/HeartsAndMinds/files/5040527/H.M_Kujari.tem_kujari.zip)

**Comment by Vdauphin** [comment]
> USAF should be already included in the rpt file.

But you didn't put it in the mods list of the issue.

**Comment by GoldJohnKing** [comment]
> > USAF should be already included in the rpt file.
> 
> But you didn't put it in the mods list of the issue.

Ah. USAF is a aircraft mod, I bet it wont affect AI behaviors so...
I'll add them to the issue mod list.

**Comment by Vdauphin** [comment]
> Ah. USAF is a aircraft mod, I bet it wont affect AI behaviors so...

RPT is asked to see error.
Mods list is asked to know what mod you use to start the pbo.
pbo is here to see what param you select and reproduce immediately the issue.

So by providing all those information I am not lossing two hours of starting your mission, see a missing mod, retarting, see an other mod missing 

If you care about mod affecting the AI behaviors, start the mission with only CBA and ACE and reproduce the issue. This will also don't waste the time of the man trying to reproduce your issue by asking him to download your mod list.

here is my paypal: https://www.paypal.me/Vdauphin

**Comment by Vdauphin** [comment]
activate/deactivate cities Can't reproduce

![20200807123020_1](https://user-images.githubusercontent.com/14364400/89638493-37c73c80-d8ac-11ea-866b-6c93317cf103.jpg)
![20200807123159_1](https://user-images.githubusercontent.com/14364400/89638496-39910000-d8ac-11ea-81ea-7770a32bd362.jpg)
![20200807124327_1](https://user-images.githubusercontent.com/14364400/89638500-3a299680-d8ac-11ea-8a4f-3421ab6587c8.jpg)
![20200807124714_1](https://user-images.githubusercontent.com/14364400/89638503-3ac22d00-d8ac-11ea-91f1-da2262913bd3.jpg)

**Comment by GoldJohnKing** [comment]
> So by providing all those information I am not lossing two hours of starting your mission, see a missing mod, retarting, see an other mod missing

You are correct, I'm really sorry for that.

> If you care about mod affecting the AI behaviors, start the mission with only CBA and ACE and reproduce the issue. This will also don't waste the time of the man trying to reproduce your issue by asking him to download your mod list.

I maintain a modded server for players, this mostly happens on production environment. At my local testing environment, I do not have many much players to activate cities, fight against enemies, and deactivate cities once and once again. As I said, the issue happens randomly. I can rarely repoduce this issue only on my testing environment with only CBA and ACE, but it did reproduce when I singleplayer for a long time with only CBA and ACE, difference is at that time, the terrain is Virolahti, not Kujari.

> here is my paypal: https://www.paypal.me/Vdauphin

I would sure to sponsor for this project if PayPal can be used in China, but as a student without visa or master card (without that means only Chinese currency can be used), the only two mobile pay software we can use are WeChat (just being banned in US) and Alipay. If you have one of them, I'm sure to sponsor for the awesome project.

> activate/deactivate cities Can't reproduce

And yes, this happens randomly. Simply activate/deactivate cities don't reproduce it.  Newly generated groups seem not reproduce this, too. I mostly notice the issue on side-mission cities that player activates, fighting, mission accomplished, and deactivates for many times.

If you think this issue is too random to be reproduced on local testing environment, I think we should close the issue, as writing a script detecting waypoints at [0,0,0] and delete them dynamically is much easier than finding the exact reason.

**Comment by Vdauphin** [comment]
> At my local testing environment, I do not have many much players to activate cities, fight against enemies, and deactivate cities once and once again. 

Just switch in debug and click on the map everywhere, you will be teleported. You are also safe because damage is disable on you

> I maintain a modded server for players, this mostly happens on production environment. 
>  As I said, the issue happens randomly. I can rarely repoduce this issue only on my testing environment with only CBA and ACE, but it did reproduce when I singleplayer for a long time with only CBA and ACE, difference is at that time, the terrain is Virolahti, not Kujari.

I don't know when you started to use the master_daily. This looks like it is now fix (by CBA 3.14 update) but you still see it because the mission is persistent so this wrong waypoint has been saved and restore each time you activate a city. You can share your save from your production environment. 

Here is what I have on my dedicated server with master_daily
![20200807151149_1](https://user-images.githubusercontent.com/14364400/89649083-79fa7900-d8c0-11ea-80b3-2c2f3344b82e.jpg)
The important is to be sure the current master_daily does not generate those wrong waypoint

> If you think this issue is too random to be reproduced on local testing environment, I think we should close the issue, as writing a script detecting waypoints at [0,0,0] and delete them dynamically is much easier than finding the exact reason.

Issue are not random when we have the reproducing step

**Comment by Vdauphin** [comment]
Just notice that if you do:
```sqf
b = createGroup east;
waypoints b apply {waypointPosition _x}
```
return `[[0,0,-0.00627938]]`... So arma create a waypoint it self even if there is no unit in the group. But for some reason, AI ignore this "ghost" waypoint, also isn't show in Zeus

The only visible issue, is that this ghost waypoint is saved between activate/deactivation of city for unit patroling around city. So this ghost waypoint is added endless at each activate/deactivation because it never removed

**Comment by GoldJohnKing** [comment]
> I don't know when you started to use the master_daily. This looks like it is now fix (by CBA 3.14 update) but you still see it because the mission is persistent so this wrong waypoint has been saved and restore each time you activate a city. You can share your save from your production environment.

When I do a manually maintenance, I manually delete saves. The save is sure to be gone every day (because I maintenance it almost every day). Even my server hold a totally clean start of the mission, the issue still occurs. (Yesterday after I submitted the issue, I cleaned the save and restarted the mission, it still occurs.) So I think it should not done with saves.

> The only visible issue, is that this ghost waypoint is saved between activate/deactivation of city for unit patroling around city. So this ghost waypoint is added endless at each activate/deactivation because it never removed

This makes sense. However the issue seems not always reproduce when I hold the mission locally and continuously activate/deactivation a city, even in a nonsense frequency, or I'll be sure to give a reproduce step.

Anyway, I'll try #938 and see if it solve the issue.

**Comment by GoldJohnKing** [comment]
Hi, I've been testing #938 for two days on my server, the issue has totally gone!
